app/src/org/commcare/views/widgets/WidgetUtils.java

```diff
@@ -33,4 +40,42 @@ public static Button setupClearButton(Context context, ViewGroup root, String te
         clearButton.setVisibility(visibility);
         return clearButton;
     }
+
+    /**
+     * Works just like {@link #createScanIntent(Context, String...)} except the format is null.
+     */
+    public static Intent createScanIntent(Context context) {
+        return createScanIntent(context, null);
+    }
+
+    /**
+     * A utility to create intent for barcode scan.
+     *
+     * If usage of 3rd party apps for scanning is enabled by the user, then this method will
+     * create a generic intent for scanning barcodes and will fallback to zxing library if no
+     * app exists to handle such intent.
+     *
+     * @param formats names of {@link com.google.zxing.BarcodeFormat}s to scan for. eg. {@link com.google.zxing.BarcodeFormat#QR_CODE}
+     */
+    public static Intent createScanIntent(Context context, String... formats) {
+        if (MainConfigurablePreferences.useIntentCalloutForScanner()) {
+            Intent scanIntent = new Intent(Intents.Scan.ACTION);
+            if (scanIntent.resolveActivity(context.getPackageManager()) != null) {
+                if (formats != null) {
+                    StringBuilder joinedByComma = new StringBuilder();
```

Comment:
can use [String.join](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#join-java.lang.CharSequence-java.lang.CharSequence...-)

Reply:
Made a change to use `TextUtils` since [String.join](https://developer.android.com/reference/java/lang/String#join(java.lang.CharSequence,%20java.lang.CharSequence...)) is added on api 26